File: PokerWorld/src/newPokerLogic/PokerGame.java
package newPokerLogic;

import newPokerLogicTools.PokerGameObserver;

public class PokerGame
{
    private final PokerGameObserver m_gameObserver;
    private final PokerTableInfo m_pokerTable;
    private TypePokerGameState m_currentGameState;
    private TypePokerGameRound m_currentGameRound;
    private TypePokerGameRoundState m_currentGameRoundState;
    
    private final AbstractPokerDealer m_pokerDealer;
    
    public PokerGame()
    {
        this(new RandomPokerDealer());
    }
    
    public PokerGame(PokerTableInfo info)
    {
        this(new RandomPokerDealer(), info);
    }
    
    public PokerGame(AbstractPokerDealer dealer)
    {
        this(new RandomPokerDealer(), new PokerTableInfo());
    }
    
    public PokerGame(AbstractPokerDealer dealer, PokerTableInfo info)
    {
        m_pokerDealer = dealer;
        m_gameObserver = new PokerGameObserver();
        m_pokerTable = info;
        m_currentGameState = TypePokerGameState.INIT;
    }
    
    public void start()
    {
        setCurrentGameState(TypePokerGameState.PLAYERS_WAITING);
    }
    
    public TypePokerGameState getCurrentGameState()
    {
        return m_currentGameState;
    }
    
    public TypePokerGameRound getCurrentGameRound()
    {
        return m_currentGameRound;
    }
    
    public TypePokerGameRoundState getCurrentGameRoundState()
    {
        return m_currentGameRoundState;
    }
    
    private void setCurrentGameState(TypePokerGameState newGS)
    {
        
        final TypePokerGameState oldGS = m_currentGameState;
        
        if (m_currentGameState == TypePokerGameState.END)
        {
            return;
        }
        
        if (newGS.ordinal() - oldGS.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameState = newGS;
        
        switch (m_currentGameState)
        {
            case INIT:
                break;
            case PLAYERS_WAITING:
                TryToBegin();
                break;
            case BLIND_WAITING:
                break;
            case PLAYING:
                m_currentGameRound = TypePokerGameRound.PREFLOP;
                m_currentGameRoundState = TypePokerGameRoundState.CARDS;
                startRound();
                break;
            case SHOWDOWN:
                showAllCards();
                break;
            case DECIDE_WINNERS:
                break;
            case MONEY_DISTRIBUTION:
                break;
            case END:
                m_gameObserver.everythingEnded(m_pokerTable);
                break;
            
        }
    }
    
    private void setCurrentGameRound(TypePokerGameRound newGR)
    {
        
        final TypePokerGameRound oldGR = m_currentGameRound;
        
        if (m_currentGameState != TypePokerGameState.PLAYING)
        {
            return;
        }
        
        if (newGR.ordinal() - oldGR.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameRoundState = TypePokerGameRoundState.CARDS;
        m_pokerTable.setCurrentPlayerNoSeat(m_pokerTable.getCurrentDealerNoSeat());
        m_currentGameRound = newGR;
        startRound();
    }
    
    private void setCurrentGameRoundState(TypePokerGameRoundState newGRS)
    {
        
        final TypePokerGameRoundState oldGRS = m_currentGameRoundState;
        
        if (m_currentGameState != TypePokerGameState.PLAYING)
        {
            return;
        }
        
        if (newGRS.ordinal() - oldGRS.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameRoundState = newGRS;
        startRound();
    }
    
    public PokerGameObserver getGameObserver()
    {
        return m_gameObserver;
    }
    
    public PokerTableInfo getPokerTable()
    {
        return m_pokerTable;
    }
    
    public boolean isRunning()
    {
        return m_currentGameState != TypePokerGameState.END;
    }
    
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // ///////////////// COMMUNICATION ///////////////////////////
    // /////////////////////////////////
    public boolean joinGame(PokerPlayerInfo p)
    {
        if (m_currentGameState == TypePokerGameState.INIT || m_currentGameState == TypePokerGameState.END)
        {
            System.err.println("Bad timing:" + m_currentGameState);
            return false;
        }
        
        return m_pokerTable.joinTable(p);
    }
    
    public void sitInGame(PokerPlayerInfo p)
    {
        m_gameObserver.playerJoined(m_pokerTable, p);
        if (m_currentGameState == TypePokerGameState.PLAYERS_WAITING)
        {
            TryToBegin();
        }
    }
    
    public boolean leaveGame(PokerPlayerInfo p)
    {
        if (m_pokerTable.leaveTable(p))
        {
            m_gameObserver.playerLeaved(m_pokerTable, p);
            return true;
        }
        return false;
    }
    
    public boolean playMoney(PokerPlayerInfo p, int amount)
    {
        final int amnt = Math.min(amount, p.getCurrentSafeMoneyAmount());
        System.out.println(p.getPlayerName() + " is playing " + amnt + " money on state: " + m_currentGameState);
        if (m_currentGameState == TypePokerGameState.BLIND_WAITING)
        {
            System.out.println(p.getPlayerName() + " is putting blind of " + amnt);
            System.out.println("Total still needed is " + m_pokerTable.getTotalBlindNeeded());
            if (amnt != m_pokerTable.blindNeeded(p))
            {
                System.err.println(p.getPlayerName() + " needed to put " + m_pokerTable.blindNeeded(p));
                return false;
            }
            if (!p.tryBet(amnt))
            {
                // TODO: authorise si ALL IN
                System.err.println(p.getPlayerName() + " just .. can't !! ");
                return false;
            }
            m_pokerTable.incTotalPotAmount(amnt);
            m_gameObserver.playerMoneyChanged(m_pokerTable, p);
            m_pokerTable.setBlindNeeded(p, 0);
            if (amnt == m_pokerTable.getSmallBlindAmount())
            {
                m_gameObserver.playerActionTaken(m_pokerTable, p, TypePokerGameAction.SMALL_BLIND_POSTED, amnt);
            }
            else
            {
                m_gameObserver.playerActionTaken(m_pokerTable, p, TypePokerGameAction.BIG_BLIND_POSTED, amnt);
            }
            m_pokerTable.setTotalBlindNeeded(m_pokerTable.getTotalBlindNeeded() - amnt);
            if (m_pokerTable.getTotalBlindNeeded() == 0)
            {
                setCurrentGameState(TypePokerGameState.PLAYING);
            }
            return true;
        }
        
        else if (m_currentGameState == TypePokerGameState.PLAYING && m_currentGameRoundState == TypePokerGameRoundState.BETTING)
        {
            System.out.println("Currently, we need " + amountNeeded(p) + " minimum money from this player");
            if (p.getCurrentTablePosition() != m_pokerTable.getCurrentPlayerNoSeat())
            {
                System.err.println("BUT SCREW YOU, IT'S NOT YOUR TURN !!!!!");
                return false;
            }
            
            if (amnt == -1)
            {
                System.out.println("So ... the little girl folds !");
                foldPlayer(p);
                continueBettingRound();
                return true;
            }
            int amntNeeded = amountNeeded(p);
            if (amnt < amntNeeded)
            {
                if (p.canBet(amnt + 1))
                {
                    System.err.println("BUT SCREW YOU, IT'S NOT ENOUGH !!!!!");
                    return false;
                }
                System.out.println("So ... All-In ! getCurrentBetMoneyAmount: " + p.getCurrentBetMoneyAmount());
                amntNeeded = amnt;
                p.setAllIn();
                m_pokerTable.incNbAllIn();
                m_pokerTable.addAllInCap(p.getCurrentBetMoneyAmount() + amnt);
                m_pokerTable.decNbPlaying();
            }
            if (!p.tryBet(amnt))
            {
                System.err.println("BUT SCREW YOU, YOU JUST CAN'T !!!!!");
                return false;
            }
            m_gameObserver.playerMoneyChanged(m_pokerTable, p);
            if (amnt == amntNeeded)
            {
                System.out.println("Will call with $" + amnt);
                m_pokerTable.incTotalPotAmount(amnt);
                callPlayer(p, amnt);
                continueBettingRound();
                return true;
            }
            System.out.println("Will raise with $" + amnt);
            m_pokerTable.incTotalPotAmount(amnt);
            raisePlayer(p, amnt);
            playNext();
            return true;
        }
        System.err.println("BUT WE DON'T CARE !!!!!");
        return false;
    }
    
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // ////// PRIV /////////////////////
    // /////////////////////////////////
    
    private void startRound()
    {
        switch (m_currentGameRoundState)
        {
            case CARDS:
                startCardRound();
                break;
            case BETTING:
                startBettingRound();
                break;
            case CUMUL:
                startCumulRound();
                break;
        }
    }
    
    private void showAllCards()
    {
        for (final PokerPlayerInfo p : m_pokerTable.getPlayers())
        {
            if (p.isPlaying() || p.isAllIn())
            {
                p.setShowingCards(true);
                m_gameObserver.playerHoleCardsChanged(m_pokerTable, p);
            }
        }
        setCurrentGameState(TypePokerGameState.DECIDE_WINNERS);
    }
    
    private void startCumulRound()
    {
        m_pokerTable.managePotsRoundEnd();
        m_gameObserver.gameBettingRoundEnded(m_pokerTable, m_currentGameRound);
        if (m_pokerTable.getNbPlaying() == 1 && m_pokerTable.getNbAllIn() == 0)
        {
            setCurrentGameState(TypePokerGameState.SHOWDOWN);
        }
        else
        {
            switch (m_currentGameRound)
            {
                case PREFLOP:
                    setCurrentGameRound(TypePokerGameRound.FLOP);
                    break;
                case FLOP:
                    setCurrentGameRound(TypePokerGameRound.TURN);
                    break;
                case TURN:
                    setCurrentGameRound(TypePokerGameRound.RIVER);
                    break;
                case RIVER:
                    setCurrentGameState(TypePokerGameState.SHOWDOWN);
                    break;
            }
        }
    }
    
    private void startBettingRound()
    {
        m_pokerTable.setNbPlayed(0);
        continueBettingRound();
    }
    
    private void startCardRound()
    {
        switch (m_currentGameRound)
        {
            case PREFLOP:
                m_pokerTable.setCurrentPlayerNoSeat(m_pokerTable.getCurrentBigBlindNoSeat());
                dealHole();
                m_pokerTable.setCurrentHigherBet(m_pokerTable.getBigBlindAmount());
                break;
            case FLOP:
                dealFlop();
                break;
            case TURN:
                dealTurn();
                break;
            case RIVER:
                dealRiver();
                break;
        }
        setCurrentGameRoundState(TypePokerGameRoundState.BETTING);
    }
    
    private void dealRiver()
    {
        m_pokerTable.addBoardCard(m_pokerDealer.dealRiver());
        m_gameObserver.gameBoardCardsChanged(m_pokerTable);
    }
    
    private void dealTurn()
    {
        m_pokerTable.addBoardCard(m_pokerDealer.dealTurn());
        m_gameObserver.gameBoardCardsChanged(m_pokerTable);
    }
    
    private void dealFlop()
    {
        m_pokerTable.addBoardCards(m_pokerDealer.dealFlop());
        m_gameObserver.gameBoardCardsChanged(m_pokerTable);
    }
    
    private void dealHole()
    {
        for (final PokerPlayerInfo p : m_pokerTable.getPlayingPlayers())
        {
            p.setHand(m_pokerDealer.dealHoles(p));
            m_gameObserver.playerHoleCardsChanged(m_pokerTable, p);
        }
    }
    
    private int amountNeeded(PokerPlayerInfo p)
    {
        return m_pokerTable.getCurrentHigherBet() - p.getCurrentBetMoneyAmount();
    }
    
    private void foldPlayer(PokerPlayerInfo p)
    {
        p.setFolded();
        m_pokerTable.decNbPlaying();
        m_gameObserver.playerActionTaken(m_pokerTable, p, TypePokerGameAction.FOLDED, -1);
    }
    
    private void TryToBegin()
    {
        System.out.print("Trying to begin ...");
        m_pokerTable.setNbPlaying(m_pokerTable.getAndSetNbPlayingPlayers());
        if (m_pokerTable.getNbPlaying() > 1)
        {
            System.out.println(" yep ! do it !");
            m_pokerTable.setNbPlayed(0);
            m_pokerTable.placeButtons();
            m_pokerTable.initPots();
            setCurrentGameState(TypePokerGameState.BLIND_WAITING);
            m_gameObserver.gameBlindsNeeded(m_pokerTable);
        }
        else
        {
            System.out.println(" just too bad :(");
            m_pokerTable.setCurrentDealerNoSeat(-1);
            m_pokerTable.setCurrentSmallBlindNoSeat(-1);
            m_pokerTable.setCurrentSmallBlindNoSeat(-1);
        }
    }
    
    private void callPlayer(PokerPlayerInfo p, int played)
    {
        m_pokerTable.incNbPlayed();
        m_gameObserver.playerActionTaken(m_pokerTable, p, TypePokerGameAction.CALLED, played);
    }
    
    private void raisePlayer(PokerPlayerInfo p, int played)
    {
        m_pokerTable.setNbPlayed(1);
        m_pokerTable.setCurrentHigherBet(p.getCurrentBetMoneyAmount());
        m_gameObserver.playerActionTaken(m_pokerTable, p, TypePokerGameAction.RAISED, played);
    }
    
    private void continueBettingRound()
    {
        if (m_pokerTable.getNbPlaying() == 1 || m_pokerTable.getNbPlayed() >= m_pokerTable.getNbPlaying())
        {
            endBettingRound();
        }
        else
        {
            playNext();
        }
    }
    
    private void endBettingRound()
    {
        setCurrentGameRoundState(TypePokerGameRoundState.CUMUL);
    }
    
    private void playNext()
    {
        final PokerPlayerInfo player = m_pokerTable.nextPlayingPlayer(m_pokerTable.getCurrentPlayerNoSeat());
        m_pokerTable.setCurrentPlayerNoSeat(player.getCurrentTablePosition());
        m_gameObserver.playerActionNeeded(m_pokerTable, player);
    }
}

File: C#/BluffinMuffin.Poker.Logic/PokerGame.cs
using System;
using System.Threading;
using BluffinMuffin.Poker.DataTypes.Enums;
using BluffinMuffin.Poker.DataTypes;
using BluffinMuffin.Poker.DataTypes.EventHandling;
using BluffinMuffin.Protocol.DataTypes;
using BluffinMuffin.Protocol.DataTypes.Enums;
using Com.Ericmas001.Util;

namespace BluffinMuffin.Poker.Logic
{
    /// <summary>
    /// This represent only one "Game" of Poker. This means from Putting Blinds to preflop, flop, turn, river and then declaring the winner.
    /// Tipically you would have many "Game" while sitting at a table, but players can sit-in and sit-out, so it's like a different game every "Game".
    /// </summary>
    public class PokerGame : IPokerGame
    {
        #region Fields

        private readonly AbstractDealer m_Dealer; // Dealer

        // STATES
        private GameStateEnum m_State; // L'etat global de la game
        private RoundStateEnum m_RoundState; // L'etat de la game pour chaque round
        #endregion Fields

        #region Properties
        public PokerGameObserver Observer { get; private set; }

        /// <summary>
        /// The Table Entity
        /// </summary>
        public TableInfo Table { get; private set; }
        /// <summary>
        /// The PokerTable Entity
        /// </summary>
        public PokerTable GameTable { get { return (PokerTable)Table; } }

        /// <summary>
        /// The Rules Entity
        /// </summary>
        public TableParams Params { get; private set; }

        /// <summary>
        /// Current Round of the Playing State
        /// </summary>
        public RoundTypeEnum Round
        {
            get { return Table.Round; }
        }

        /// <summary>
        /// Is the Game currently Running ? (Not Ended)
        /// </summary>
        public bool IsRunning
        {
            get { return m_State != GameStateEnum.End; }
        }

        /// <summary>
        /// Is the Game currently Running ? (Not Ended)
        /// </summary>
        public bool IsPlaying
        {
            get { return IsRunning && m_State >= GameStateEnum.WaitForBlinds; }
        }

        /// <summary>
        /// Current State of the Game
        /// </summary>
        public GameStateEnum State
        {
            get { return m_State; }
        }

        private int m_NbGamePlayed = 0;

        #endregion

        #region Ctors & Init

        public PokerGame(PokerTable table)
            : this(new TexasHoldemDealer(), table)
        {
        }

        private PokerGame(AbstractDealer dealer, PokerTable table)
        {
            Observer = new PokerGameObserver(this);
            m_Dealer = dealer;
            Table = table;
            Params = table.Params;
            m_State = GameStateEnum.Init;
            m_RoundState = RoundStateEnum.Cards;
        }
        #endregion Ctors & Init

        #region Public Methods
        /// <summary>
        /// Starts the Game. Only works in Init State
        /// </summary>
        public void Start()
        {
            if( m_State == GameStateEnum.Init )
                AdvanceToNextGameState(); //Advancing to WaitForPlayers State
        }

        /// <summary>
        /// Add a player to the table
        /// </summary>
        public bool JoinGame(PlayerInfo p)
        {
            if (m_State == GameStateEnum.Init || m_State == GameStateEnum.End)
            {
                LogManager.Log(LogLevel.Error, "PokerGame.JoinGame", "Can't join, bad timing: {0}", m_State);
                return false;
            }

            Observer.RaisePlayerJoined(p);
            return GameTable.JoinTable(p);
        }

        public int AfterPlayerSat(PlayerInfo p, int noSeat = -1, int moneyAmount = 1500)
        {
            var seat = p.NoSeat == -1 ? null : Table.Seats[p.NoSeat];
            if (seat != null && !seat.IsEmpty)
            {
                Observer.RaiseSeatUpdated(seat.Clone());

                if (m_State == GameStateEnum.WaitForPlayers)
                    TryToBegin();
                else if (m_State > GameStateEnum.WaitForPlayers)
                    GameTable.NewArrivals.Add(p);
                return p.NoSeat;
            }
            return -1;
        }

        public bool SitOut(PlayerInfo p)
        {
            var oldSeat = p.NoSeat;
            if (oldSeat == -1)
                return true;

            var blindNeeded = GameTable.GetBlindNeeded(p);

            p.State = PlayerStateEnum.Zombie;
            if (State == GameStateEnum.Playing && Table.CurrentPlayer == p)
                PlayMoney(p, -1);
            else if (blindNeeded > 0)
                PlayMoney(p, blindNeeded);

            if (Table.SeatsContainsPlayer(p) && Table.SitOut(p))
            {
                var seat = new SeatInfo()
                {
                    Player = null,
                    NoSeat = oldSeat,
                };
                Observer.RaiseSeatUpdated(seat);
                TryToBegin();
                return true;
            }
            return false;
        }

        /// <summary>
        /// The player is leaving the game
        /// </summary>
        public void LeaveGame(PlayerInfo p)
        {
            var sitOutOk = SitOut(p);

            if (sitOutOk && Table.LeaveTable(p))
            {
                if (Table.Players.Count == 0)
                    m_State = GameStateEnum.End;
            }
        }

        /// <summary>
        /// The player is putting money in the game
        /// </summary>
        public bool PlayMoney(PlayerInfo p, int amount)
        {
            lock(Table)
            {
                var amnt = Math.Min(amount, p.MoneySafeAmnt);
                LogManager.Log(LogLevel.MessageLow, "PokerGame.PlayMoney", "{0} is playing {1} money on state: {2}", p.Name, amnt, m_State);

                if (m_State == GameStateEnum.WaitForBlinds)
                    return PlayBlinds(p, amnt);
                if (m_State == GameStateEnum.Playing && m_RoundState == RoundStateEnum.Betting)
                    return BetMoney(p, amnt);

                LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} played money but the game is not in the right state", p.Name);

                return false;
            }
        }
        #endregion

        #region Private Methods
        private void AdvanceToNextGameState()
        {
            if (m_State == GameStateEnum.End)
                return;

            m_State = (GameStateEnum)(((int)m_State) + 1);

            switch (m_State)
            {
                case GameStateEnum.Init:
                    break;
                case GameStateEnum.WaitForPlayers:
                    TryToBegin();
                    break;
                case GameStateEnum.WaitForBlinds:
                    Table.HigherBet = 0;
                    //If we got all the blinds, what are we waiting for ?!
                    if (GameTable.TotalBlindNeeded == 0)
                        AdvanceToNextGameState(); //Advancing to Playing State
                    break;
                case GameStateEnum.Playing:
                    Table.Round = RoundTypeEnum.Preflop;
                    m_RoundState = RoundStateEnum.Cards;
                    StartRound();
                    break;
                case GameStateEnum.Showdown:
                    ShowAllCards();
                    break;
                case GameStateEnum.DecideWinners:
                    DecideWinners();
                    break;
                case GameStateEnum.DistributeMoney:
                    DistributeMoney();
                    StartANewGame();
                    break;
                case GameStateEnum.End:
                    Observer.RaiseEverythingEnded();
                    break;
            }
        }
        private void StartANewGame()
        {
            Observer.RaiseGameEnded();
            m_NbGamePlayed++;
            m_State = GameStateEnum.WaitForPlayers;
            TryToBegin();
        }
        private void AdvanceToNextRound()
        {
            if (m_State != GameStateEnum.Playing)
                return;

            if (Table.Round == RoundTypeEnum.River)
                AdvanceToNextGameState(); //Advancing to Showdown State
            else
            {
                m_RoundState = RoundStateEnum.Cards;
                Table.ChangeCurrentPlayerTo(Table.DealerSeat);
                Table.Round = (RoundTypeEnum)(((int)Table.Round) + 1);
                StartRound();
            }
        }
        private void AdvanceToNextRoundState()
        {
            if (m_State != GameStateEnum.Playing)
                return;

            if (m_RoundState == RoundStateEnum.Cumul)
                return;

            m_RoundState = (RoundStateEnum)(((int)m_RoundState) + 1);
            StartRound();
        }
        private bool BetMoney(PlayerInfo p, int amnt)
        {
            LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "Currently, we need {0} minimum money from this player", Table.CallAmnt(p));

            //Validation: Is it the player's turn to play ?
            if (p.NoSeat != Table.NoSeatCurrentPlayer)
            {
                LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} just played but it wasn't his turn", p.Name); 
                return false;
            }

            //The Player is Folding
            if (amnt == -1)
            {
                LogManager.Log(LogLevel.MessageLow, "PokerGame.PlayMoney", "{0} FOLDED", p.Name);
                FoldPlayer(p);
                ContinueBettingRound();
                return true;
            }

            var amntNeeded = Table.CallAmnt(p);

            //Validation: Is the player betting under what he needs to Call ?
            if (amnt < amntNeeded)
            {
                //Validation: Can the player bet more ? If yes, this is not fair.
                if (p.CanBet(amnt + 1))
                {
                    LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} needed to play at least {1} and tried {2}", p.Name, amntNeeded, amnt);
                    return false;
                }

                //Else, well, that's ok, the player is All-In !
                amntNeeded = amnt;
            }

            if (amnt > amntNeeded && amnt < Table.MinRaiseAmnt(p)) 
            {
                LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} needed to play at least {1} to raise and tried {2}", p.Name, amntNeeded, amnt);
                return false;
            }

            //Let's hope the player has enough money ! Time to Bet!
            if (!p.TryBet(amnt))
            {
                LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} just put more money than he actually have ({1} > {2})", p.Name, amnt, p.MoneySafeAmnt);
                return false;
            }

            //Update the MinimumRaiseAmount
            Table.MinimumRaiseAmount = Math.Max(Table.MinimumRaiseAmount, p.MoneyBetAmnt);

            //Notify the change in the player's account
            Observer.RaisePlayerMoneyChanged(p);

            //Is the player All-In?
            if (p.MoneySafeAmnt == 0)
            {
                LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "Player now All-In !");
                p.State = PlayerStateEnum.AllIn;
                Table.NbAllIn++;
                GameTable.AddAllInCap(p.MoneyBetAmnt);
            }

            //Hmmm ... More Money !! 
            Table.TotalPotAmnt += amnt;

            //Was it a CALL or a RAISE ?
            if (amnt == amntNeeded)
            {
                LogManager.Log(LogLevel.MessageLow, "PokerGame.PlayMoney", "{0} CALLED WITH ${1}", p.Name, amnt);
                CallPlayer(p, amnt);
            }
            else
            {
                LogManager.Log(LogLevel.MessageLow, "PokerGame.PlayMoney", "{0} RAISED WITH ${1}", p.Name, amnt);
                RaisePlayer(p, amnt);
            }

            // Ok this player received enough attention !
            ContinueBettingRound();

            return true;
        }
        private bool PlayBlinds(PlayerInfo p, int amnt)
        {
            LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "Total blinds needed is {0}", GameTable.TotalBlindNeeded);
            LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "{0} is putting blind of {1}", p.Name, amnt);

            //What is the need Blind from the player ?
            var needed = GameTable.GetBlindNeeded(p);

            //If the player isn't giving what we expected from him
            if (amnt != needed)
            {
                //If the player isn't playing enough but it's all he got, time to go All-In
                if (amnt < needed && !p.CanBet(amnt + 1))
                {
                    LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "Player now All-In !");
                    p.State = PlayerStateEnum.AllIn;
                    Table.NbAllIn++;
                    GameTable.AddAllInCap(p.MoneyBetAmnt + amnt);
                }
                else //well, it's just not fair to play that
                {
                    LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} needed to put a blind of {1} and tried {2}", p.Name, needed, amnt);
                    return false;
                }
            }

            //Let's hope the player has enough money ! Time to put the blinds !
            if (!p.TryBet(amnt))
            {
                LogManager.Log(LogLevel.Warning, "PokerGame.PlayMoney", "{0} just put more money than he actually have ({1} > {2})", p.Name, amnt, p.MoneySafeAmnt);
                return false;
            }

            //Hmmm ... More Money !! 
            Table.TotalPotAmnt += amnt;

            //Notify the change in the player's account
            Observer.RaisePlayerMoneyChanged(p);

            //Take note of the given Blind Amount for the player.
            GameTable.SetBlindNeeded(p, 0);

            //Take note of the action
            var whatAmIDoing = GameActionEnum.PostAnte;
            if(Table.Params.Blind.OptionType == BlindTypeEnum.Blinds)
            {
                var bob = Table.Params.Blind as BlindOptionsBlinds;
                if (bob != null && needed == bob.SmallBlindAmount)
                    whatAmIDoing = GameActionEnum.PostSmallBlind;
                else
                    whatAmIDoing = GameActionEnum.PostBigBlind;
            }
            LogManager.Log(LogLevel.MessageLow, "PokerGame.PlayMoney", "{0} POSTED BLIND ({1})", p.Name, whatAmIDoing);
            Observer.RaisePlayerActionTaken(p, whatAmIDoing, amnt);

            //Let's set the HigherBet
            if (amnt > Table.HigherBet)
                Table.HigherBet = amnt;

            LogManager.Log(LogLevel.MessageVeryLow, "PokerGame.PlayMoney", "Total blinds still needed is {0}", GameTable.TotalBlindNeeded);

            //If we got all the blinds, what are we waiting for ?!
            if (GameTable.TotalBlindNeeded == 0)
                AdvanceToNextGameState(); //Advancing to Playing State
            return true;
        }
        private void StartRound()
        {
            switch (m_RoundState)
            {
                case RoundStateEnum.Cards:
                    StartCardRound();
                    break;
                case RoundStateEnum.Betting:
                    StartBettingRound();
                    break;
                case RoundStateEnum.Cumul:
                    StartCumulRound();
                    break;
            }
        }
        private void StartCumulRound()
        {
            GameTable.ManagePotsRoundEnd();

            Observer.RaiseGameBettingRoundEnded(Table.Round);

            if (Table.NbPlayingAndAllIn <= 1)
                AdvanceToNextGameState(); //Advancing to Showdown State
            else
                AdvanceToNextRound(); //Advancing to Next Round
        }
        private void StartBettingRound()
        {
            Observer.RaiseGameBettingRoundStarted(Table.Round);

            //We Put the current player just before the starting player, then we will take the next player and he will be the first
            Table.ChangeCurrentPlayerTo(Table.GetSeatOfPlayingPlayerJustBefore(Table.SeatOfTheFirstPlayer));
            Table.NbPlayed = 0;
            Table.MinimumRaiseAmount = Table.Params.MoneyUnit;

            WaitALittle(Params.WaitingTimes.AfterBoardDealed);

            if (Table.NbPlaying <= 1)
                EndBettingRound();
            else
                ContinueBettingRound();
        }
        private void StartCardRound()
        {
            switch (Table.Round)
            {
                case RoundTypeEnum.Preflop:
                    DealHole();
                    break;
                case RoundTypeEnum.Flop:
                    DealFlop();
                    break;
                case RoundTypeEnum.Turn:
                    DealTurn();
                    break;
                case RoundTypeEnum.River:
                    DealRiver();
                    break;
            }

            AdvanceToNextRoundState(); // Advance to Betting Round State
        }
        private void DealRiver()
        {
            GameTable.AddCards(m_Dealer.DealRiver());
        }
        private void DealTurn()
        {
            GameTable.AddCards(m_Dealer.DealTurn());
        }
        private void DealFlop()
        {
            GameTable.AddCards(m_Dealer.DealFlop());
        }
        private void DealHole()
        {
            foreach (var p in Table.PlayingAndAllInPlayers)
            {
                p.Cards = m_Dealer.DealHoles();
                Observer.RaisePlayerHoleCardsChanged(p);
            }
        }
        private void ShowAllCards()
        {
            foreach (var p in Table.Players)
                if (p.IsPlaying || p.IsAllIn)
                {
                    p.IsShowingCards = true;
                    Observer.RaisePlayerHoleCardsChanged(p);
                }
            AdvanceToNextGameState(); //Advancing to DecideWinners State
        }
        private void FoldPlayer(PlayerInfo p)
        {
            if(p.State != PlayerStateEnum.Zombie)
                p.State = PlayerStateEnum.SitIn;

            WaitALittle(Params.WaitingTimes.AfterPlayerAction);

            Observer.RaisePlayerActionTaken(p, GameActionEnum.Fold, -1);
        }
        private void CallPlayer(PlayerInfo p, int played)
        {
            Table.NbPlayed++;

            WaitALittle(Params.WaitingTimes.AfterPlayerAction);

            Observer.RaisePlayerActionTaken(p, GameActionEnum.Call, played);
        }
        private void RaisePlayer(PlayerInfo p, int played)
        {
            // Since every raise "restart" the round, 
            // the number of players who played is the number of AllIn players plus the raising player
            Table.NbPlayed = Table.NbAllIn;
            if (!p.IsAllIn)
                Table.NbPlayed++;

            Table.HigherBet = p.MoneyBetAmnt;

            WaitALittle(Params.WaitingTimes.AfterPlayerAction);

            Observer.RaisePlayerActionTaken(p, GameActionEnum.Raise, played);
        }
        private void ContinueBettingRound()
        {
            if (Table.NbPlayingAndAllIn == 1 || Table.NbPlayed >= Table.NbPlayingAndAllIn)
                EndBettingRound();
            else
                ChooseNextPlayer();
        }
        private void EndBettingRound()
        {
            AdvanceToNextRoundState(); // Advance to Cumul Round State
        }
        private void ChooseNextPlayer()
        {
            var next = Table.GetSeatOfPlayingPlayerNextTo(Table.CurrentPlayerSeat);

            Table.ChangeCurrentPlayerTo(next);

            Observer.RaisePlayerActionNeeded(next.Player);

            if (next.Player.IsZombie)
            {
                if (Table.CanCheck(next.Player))
                    PlayMoney(next.Player, 0);
                else
                    PlayMoney(next.Player, -1);
            }
        }
        private void DistributeMoney()
        {
            foreach (var pot in Table.Pots)
            {
                var players = pot.AttachedPlayers;
                if (players.Length > 0)
                {
                    var wonAmount = pot.Amount / players.Length;
                    if (wonAmount > 0)
                    {
                        foreach (var p in players)
                        {
                            p.MoneySafeAmnt += wonAmount;
                            Observer.RaisePlayerMoneyChanged(p);
                            Observer.RaisePlayerWonPot(p, pot.Id, wonAmount);
                            WaitALittle(Params.WaitingTimes.AfterPotWon);
                        }
                    }
                }
            }
        }
        private void DecideWinners()
        {
            GameTable.CleanPotsForWinning();
            AdvanceToNextGameState(); //Advancing to DistributeMoney State
        }
        private void WaitALittle(int waitingTime)
        {
            Thread.Sleep(waitingTime);
        }
        private void TryToBegin()
        {
            foreach (var p in Table.Players)
            {
                if (p.IsZombie)
                    LeaveGame(p);
                else if (p.CanPlay)
                {
                    p.State = PlayerStateEnum.Playing;
                    p.IsShowingCards = false;
                }
                else
                    p.State = PlayerStateEnum.SitIn;
            }
            if (Table.NbPlaying == 0)
                Observer.RaiseEverythingEnded();
            else if (Table.NbPlaying >= Table.Params.MinPlayersToStart)
            {
                Table.Params.MinPlayersToStart = 2;
                Table.InitTable();
                m_Dealer.FreshDeck();
                AdvanceToNextGameState(); //Advancing to WaitForBlinds State
                Observer.RaiseGameGenerallyUpdated();
                Observer.RaiseGameBlindNeeded();
            }
            else
            {
                if (Table.DealerSeat != null)
                    Table.DealerSeat.Attributes.Remove(SeatAttributeEnum.Dealer);
                Table.PlayingPlayers.ForEach(x => x.State = PlayerStateEnum.SitIn);
            }
        }
        #endregion Private Methods
    }
}
